File: week2/MarshalLee/Stack.cs
using System;

public class Stack<T>
{
    public int Length;
    private Node<T>? Head;

    public Stack()
    {
        this.Head = null;
        this.Length = 0;
    }

    public void Push(T data)
    {
        Node<T> node = new(data);
        this.Length++;

        if (this.Head is null)
        {
            this.Head = node;
            return;
        }

        node.Prev = this.Head;
        this.Head = node;
    }

    public T? Pop()
    {
        this.Length = Math.Max(0, this.Length - 1);

        if (this.Length == 0)
        {
            Node<T>? head0 = this.Head as Node<T>;
            this.Head = null;
            return head0.Data;
        }

        Node<T>? head = this.Head as Node<T>;
        this.Head = head.Prev;

        return head.Data;
    }

    public T? Peek<T>() where T : class
    {
        return this.Head is not null ? this.Head.Data as T : null;
    }
}

File: week2/Eren-Taskiran/Stack.java
public class Stack {
    private Node<Integer> head;
    private int size;

    public void push(int value){
        head = new Node<Integer>(value, head);
        size++;
    }

    public int pop(){
        if(isEmpty()){
            throw new IllegalStateException("Stack is empty");
        }
        int value=head.value;
        head=head.prev;
        size--;
        return value;
    }

    public int peek(){
        if(isEmpty()){
            throw new IllegalStateException("Stack is empty");
        }
        return head.value;
    }

    public int size(){
        return size;
    }

    public boolean isEmpty(){
        return size==0;
    }

    @Override
    public String toString(){
        StringBuilder sb = new StringBuilder();
        Node<Integer> current=head;

        if(current!=null){
            sb.append(current.value);
            current=current.prev;
        }

        while (current!=null){
            sb.append(" -> ");
            sb.append(current.value);
            current=current.prev;
        }

        return sb.toString();
    }
}
class Node<T>{
    public T value;
    public Node<T> prev;

    public Node(T value){
        this.value=value;
        this.prev=null;
    }
    public Node(T value,Node<T> prev){
        this.value=value;
        this.prev=prev;
    }
}